app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java

```diff
@@ -215,53 +208,112 @@ protected void attachBaseContext(Context base) {
     }
 
-    protected void updateTempState(LoyaltyCardField fieldName, Object value) {
-        tempLoyaltyCard = updateTempState(tempLoyaltyCard, fieldName, value);
+    protected void setLoyaltyCardNote(String note) {
+        tempLoyaltyCard.setNote(note);
 
-        if (initDone && (fieldName == LoyaltyCardField.cardId || fieldName == LoyaltyCardField.barcodeId || fieldName == LoyaltyCardField.barcodeType)) {
-            generateBarcode();
-        }
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardBalance(BigDecimal balance) {
+        tempLoyaltyCard.setBalance(balance);
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardBalanceType(Currency balanceType) {
+        tempLoyaltyCard.setBalanceType(balanceType);
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardCardId(String cardId) {
+        tempLoyaltyCard.setCardId(cardId);
+
+        generateBarcode();
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardBarcodeId(String barcodeId) {
+        tempLoyaltyCard.setBarcodeId(barcodeId);
+
+        generateBarcode();
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardBarcodeType(CatimaBarcode barcodeType) {
+        tempLoyaltyCard.setBarcodeType(barcodeType);
+
+        generateBarcode();
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardHeaderColor(Integer headerColor) {
+        tempLoyaltyCard.setHeaderColor(headerColor);
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardValidFrom(Date validFrom) {
+        tempLoyaltyCard.setValidFrom(validFrom);
+
+        hasChanged = true;
+    }
+
+    protected void setLoyaltyCardExpiry(Date expiry) {
+        tempLoyaltyCard.setExpiry(expiry);
 
         hasChanged = true;
     }
 
-    private void extractIntentFields(Intent intent) {
+    /* Extract intent fields and return if code should keep running */
+    private boolean extractIntentFields(Intent intent) {
         final Bundle b = intent.getExtras();
+
+        addGroup = b != null ? b.getString(BUNDLE_ADDGROUP) : null;
+        openSetIconMenu = b != null && b.getBoolean(BUNDLE_OPEN_SET_ICON_MENU, false);
+
         loyaltyCardId = b != null ? b.getInt(BUNDLE_ID) : 0;
         updateLoyaltyCard = b != null && b.getBoolean(BUNDLE_UPDATE, false);
         duplicateFromLoyaltyCardId = b != null && b.getBoolean(BUNDLE_DUPLICATE_ID, false);
+        importLoyaltyCardUri = intent.getData();
 
-        openSetIconMenu = b != null && b.getBoolean(BUNDLE_OPEN_SET_ICON_MENU, false);
-
-        cardId = b != null ? b.getString(BUNDLE_CARDID) : null;
-        barcodeId = b != null ? b.getString(BUNDLE_BARCODEID) : null;
-        barcodeType = b != null ? b.getString(BUNDLE_BARCODETYPE) : null;
-        addGroup = b != null ? b.getString(BUNDLE_ADDGROUP) : null;
+        // If we have to import a loyalty card, do so
+        if (updateLoyaltyCard || duplicateFromLoyaltyCardId) {
```

Comment:
The whole lifecycle of the edit activity is a mess, yeah.

Basically, what it used to do was store the new card data in temporary variables, then on the first `onResume` read those variables and then clear the variables. It was messy but kinda worked. This way the attempt to import is only ran on activity launch, when the activity is started with a bundle. This should be a more correct version: `execute once` instead of `execute each load (also "alt-tabbing away from the app") and wipe the temporary variable to prevent multiple executions`